File: extract_viz_slices.jl
# Extract 6 2D slices from the latest 1/24° continued checkpoint into a small
# native JLD2 file (Julia-friendly serialization).
#
# Run: ITER=9000 julia --project=. extract_viz_slices.jl

using JLD2

iter_num = parse(Int, get(ENV, "ITER", "9000"))
src = joinpath(@__DIR__, "simulations", "initial_conditions",
               "twentyfourth_continued_iter$(iter_num)_assembled.jld2")
dst = joinpath(@__DIR__, "viz_slices_iter$(iter_num).jld2")
isfile(src) || error("Missing source: $src")

k_mid = 15           # ~7 km (Δz = 30000/64 = 468.75 m → z(k=15) ≈ 7.27 km)
Δz    = 30000.0 / 64.0

@info "Reading" src k_mid

ρ_sfc, ρu_sfc, ρv_sfc_full, ρqv_sfc, ρ_mid, ρqcl_mid, ρw_mid, ρqr_volume =
    JLD2.jldopen(src, "r") do f
        Nx, Ny, Nz = size(f["ρ"])
        @info "Grid" Nx Ny Nz
        (f["ρ"][:, :, 1],                   # surface ρ
         f["ρu"][:, :, 1],                  # Periodic x-Face → same (Nx, Ny)
         f["ρv"][:, :, 1],                  # Face-y Bounded → (Nx, Ny+1)
         f["ρqᵛ"][:, :, 1],                 # (Nx, Ny)
         f["ρ"][:, :, k_mid + 1],           # mid-level ρ (Julia 1-based, so index k_mid+1)
         f["micro_ρqᶜˡ"][:, :, k_mid + 1],  # mid-level cloud liquid
         f["ρw"][:, :, k_mid + 1],          # Face-z, mid-level
         f["ρqʳ"])                          # full 3D for column integration
    end

Nx, Ny = size(ρ_sfc)
ρv_sfc = ρv_sfc_full[:, 1:Ny]  # drop the wall face row

safe_sfc = max.(ρ_sfc, 0.001f0)
safe_mid = max.(ρ_mid, 0.001f0)

u_sfc   = Float32.(ρu_sfc  ./ safe_sfc)
v_sfc   = Float32.(ρv_sfc  ./ safe_sfc)
qv_sfc  = Float32.(ρqv_sfc ./ safe_sfc .* 1000f0)     # g/kg
qcl_mid = Float32.(ρqcl_mid ./ safe_mid .* 1000f0)    # g/kg
w_mid   = Float32.(ρw_mid  ./ safe_mid)               # m/s

# Column-integrated rain: Σ_k ρqʳ[:,:,k] * Δz → kg/m²
@info "Integrating ρqʳ vertically…"
qr_vertint = Float32.(dropdims(sum(ρqr_volume; dims=3); dims=3) .* Δz)  # (Nx, Ny)

lat = Float32.(range(-80, 80; length=Ny + 1)[1:Ny] .+ (160.0f0 / Ny / 2))  # cell centers
lon = Float32.(range(0, 360; length=Nx + 1)[1:Nx] .+ (360.0f0 / Nx / 2))

@info "Writing" dst
JLD2.jldopen(dst, "w") do f
    f["iteration"]   = iter_num
    f["sim_time_h"]  = iter_num * 0.8 / 3600
    f["k_mid"]       = k_mid
    f["z_mid_km"]    = (k_mid + 0.5) * Δz / 1000
    f["dz_m"]        = Δz
    f["source"]      = basename(src)
    f["lon_deg"]     = lon
    f["lat_deg"]     = lat
    f["u_sfc"]       = u_sfc
    f["v_sfc"]       = v_sfc
    f["qv_sfc"]      = qv_sfc
    f["qr_vertint"]  = qr_vertint
    f["qcl_mid"]     = qcl_mid
    f["w_mid"]       = w_mid
end

println("  u_sfc   range: [", extrema(u_sfc), "]")
println("  v_sfc   range: [", extrema(v_sfc), "]")
println("  qv_sfc  range: [", extrema(qv_sfc), "]  (g/kg)")
println("  qr_vertint range: [", extrema(qr_vertint), "]  (kg/m²)")
println("  qcl_mid range: [", extrema(qcl_mid), "]  (g/kg)")
println("  w_mid   range: [", extrema(w_mid), "]  (m/s)")

sz = filesize(dst) / 1e6
@info "Saved" dst round(sz; digits=1)

File: extract_viz_slices.py
"""Extract 6 2D slices from the latest 1/24° continued checkpoint into a
small JLD2/HDF5 file for remote visualization.

Usage: ITER=9000 python3 extract_viz_slices.py
"""
import os, sys
import numpy as np
import h5py

iter_num = int(os.environ.get("ITER", "9000"))
src = f"/teamspace/studios/this_studio/GB-25/simulations/initial_conditions/twentyfourth_continued_iter{iter_num}_assembled.jld2"
dst = f"/teamspace/studios/this_studio/GB-25/viz_slices_iter{iter_num}.h5"

if not os.path.isfile(src):
    print(f"ERROR: missing {src}", file=sys.stderr); sys.exit(1)

k_mid = 15      # ~7 km (Δz = 30000/64 = 468.75 m → z(k=15) ≈ 7.27 km)
Δz = 30000.0 / 64.0  # m

print(f"Reading {src}")
with h5py.File(src, "r") as f:
    # h5py sees Julia's column-major as reversed: shape (Nz, Ny, Nx)
    # So f["ρ"][0] gives the k=0 (surface) slice with shape (Ny, Nx).
    # We transpose each slice to (Nx, Ny) for Julia-natural row-major analysis.
    Nz, Ny, Nx = f["ρ"].shape
    print(f"  grid: Nx={Nx}, Ny={Ny}, Nz={Nz}")

    rho_sfc  = np.array(f["ρ"][0]).T                 # (Nx, Ny)
    rho_mid  = np.array(f["ρ"][k_mid]).T

    # Momentum at surface
    ru_sfc   = np.array(f["ρu"][0]).T                # Periodic x-Face, same (Nx, Ny)
    # ρv is (Nz, Ny+1, Nx) — trim the bottom boundary row so we have (Nx, Ny)
    # h5py sees it as (Nz, Ny+1, Nx), taking [0, :Ny, :] gives (Ny, Nx)
    rv_sfc_full = np.array(f["ρv"][0])               # (Ny+1, Nx)
    rv_sfc   = rv_sfc_full[:Ny, :].T                 # (Nx, Ny)

    # Moisture
    rqv_sfc  = np.array(f["ρqᵛ"][0]).T               # (Nx, Ny) — centers
    rqcl_mid = np.array(f["micro_ρqᶜˡ"][k_mid]).T    # (Nx, Ny)

    # ρw has Nz+1 z-face levels; interior k-index for mid-level = k_mid
    # f["ρw"].shape = (Nz+1, Ny, Nx)
    rw_mid   = np.array(f["ρw"][k_mid]).T            # (Nx, Ny)

    # Vertically integrated rain: Σ ρqʳ * Δz  →  kg/m² (column mass)
    # Read entire ρqʳ volume — 8640×3840×64 Float32 = 8.5 GB
    print("  loading ρqʳ full volume for column integration…")
    rqr = np.array(f["ρqʳ"])                          # (Nz, Ny, Nx)
    qr_vertint = (rqr.sum(axis=0) * Δz).T             # (Nx, Ny), kg/m²
    del rqr

# Derived (per-mass) fields
safe_sfc = np.where(rho_sfc > 0.001, rho_sfc, 0.001)
safe_mid = np.where(rho_mid > 0.001, rho_mid, 0.001)

u_sfc   = ru_sfc  / safe_sfc
v_sfc   = rv_sfc  / safe_sfc
qv_sfc  = rqv_sfc / safe_sfc * 1000.0                 # g/kg
qcl_mid = rqcl_mid / safe_mid * 1000.0                # g/kg
w_mid   = rw_mid  / safe_mid                           # m/s

lat = np.linspace(-80.0, 80.0, Ny, endpoint=False, dtype=np.float32)
lon = np.linspace(0.0, 360.0, Nx, endpoint=False, dtype=np.float32)

print(f"Writing {dst}")
with h5py.File(dst, "w") as f:
    f.attrs["iteration"] = iter_num
    f.attrs["sim_time_h"] = iter_num * 0.8 / 3600.0
    f.attrs["k_mid"] = k_mid
    f.attrs["z_mid_km"] = (k_mid + 0.5) * Δz / 1000.0
    f.attrs["dz_m"] = Δz
    f.attrs["source"] = os.path.basename(src)
    f.create_dataset("lon_deg", data=lon, compression="gzip", compression_opts=4)
    f.create_dataset("lat_deg", data=lat, compression="gzip", compression_opts=4)
    f.create_dataset("u_sfc",   data=u_sfc.astype(np.float32),   compression="gzip", compression_opts=4)
    f.create_dataset("v_sfc",   data=v_sfc.astype(np.float32),   compression="gzip", compression_opts=4)
    f.create_dataset("qv_sfc",  data=qv_sfc.astype(np.float32),  compression="gzip", compression_opts=4)
    f.create_dataset("qr_vertint", data=qr_vertint.astype(np.float32), compression="gzip", compression_opts=4)
    f.create_dataset("qcl_mid", data=qcl_mid.astype(np.float32), compression="gzip", compression_opts=4)
    f.create_dataset("w_mid",   data=w_mid.astype(np.float32),   compression="gzip", compression_opts=4)

print(f"  u_sfc   range: [{u_sfc.min():.2f}, {u_sfc.max():.2f}] m/s")
print(f"  v_sfc   range: [{v_sfc.min():.2f}, {v_sfc.max():.2f}] m/s")
print(f"  qv_sfc  range: [{qv_sfc.min():.3f}, {qv_sfc.max():.3f}] g/kg")
print(f"  qr_vertint range: [{qr_vertint.min():.4f}, {qr_vertint.max():.4f}] kg/m²")
print(f"  qcl_mid range: [{qcl_mid.min():.3f}, {qcl_mid.max():.3f}] g/kg")
print(f"  w_mid   range: [{w_mid.min():.2f}, {w_mid.max():.2f}] m/s")

sz = os.path.getsize(dst) / 1e6
print(f"Saved {dst} ({sz:.1f} MB)")
